src/main/drivers/compass/compass_lis2mdl.c

```diff
@@ -28,114 +28,78 @@
-#define LIS2MDL_INT_CTRL_REG_XIEN           0x80
-#define LIS2MDL_INT_CTRL_REG_YIEN           0x40
-#define LIS2MDL_INT_CTRL_REG_ZIEN           0x20
-#define LIS2MDL_INT_CTRL_REG_IEA            0x04
-#define LIS2MDL_INT_CTRL_REG_IEL            0x02
-#define LIS2MDL_INT_CTRL_REG_IEN            0x01
-
-#define LIS2MDL_INT_SOURCE_REG      0x64
-#define LIS2MDL_INT_SOURCE_REG_P_TH_S_X     0x80
-#define LIS2MDL_INT_SOURCE_REG_P_TH_S_Y     0x40
-#define LIS2MDL_INT_SOURCE_REG_P_TH_S_Z     0x20
-#define LIS2MDL_INT_SOURCE_REG_N_TH_S_X     0x10
-#define LIS2MDL_INT_SOURCE_REG_N_TH_S_Y     0x08
-#define LIS2MDL_INT_SOURCE_REG_N_TH_S_Z     0x04
-#define LIS2MDL_INT_SOURCE_REG_MROI         0x02
-#define LIS2MDL_INT_SOURCE_REG_INT          0x01
-
-#define LIS2MDL_INT_THS_L_REG       0x65
-#define LIS2MDL_INT_THS_H_REG       0x66
-
-#define LIS2MDL_STATUS_REG          0x67
-#define LIS2MDL_STATUS_REG_ZXYOR            0x80
-#define LIS2MDL_STATUS_REG_ZOR              0x40
-#define LIS2MDL_STATUS_REG_YOR              0x20
-#define LIS2MDL_STATUS_REG_XOR              0x10
-#define LIS2MDL_STATUS_REG_ZXYDA            0x08
-#define LIS2MDL_STATUS_REG_ZDA              0x04
-#define LIS2MDL_STATUS_REG_YDA              0x02
-#define LIS2MDL_STATUS_REG_XDA              0x01
-
-#define LIS2MDL_OUTX_L_REG          0x68
-#define LIS2MDL_OUTX_H_REG          0x69
-#define LIS2MDL_OUTY_L_REG          0x6A
-#define LIS2MDL_OUTY_H_REG          0x6B
-#define LIS2MDL_OUTZ_L_REG          0x6C
-#define LIS2MDL_OUTZ_H_REG          0x6D
-
-#define LIS2MDL_TEMP_OUT_L_REG      0x6E
-#define LIS2MDL_TEMP_OUT_H_REG      0x6F
-
-static bool lis2mdlRead(magDev_t * mag, int16_t *magData)
+#define LIS2MDL_MAG_I2C_ADDRESS 0x1E
+
+// LIS2MDL Registers
+#define LIS2MDL_ADDR_CFG_REG_A  0x60
+#define LIS2MDL_ADDR_CFG_REG_B  0x61
+#define LIS2MDL_ADDR_CFG_REG_C  0x62
+#define LIS2MDL_ADDR_STATUS_REG 0x67
+#define LIS2MDL_ADDR_OUTX_L_REG 0x68
+#define LIS2MDL_ADDR_WHO_AM_I   0x4F
+
+// LIS2MDL Definitions
+#define LIS2MDL_WHO_AM_I         0b01000000
+#define LIS2MDL_STATUS_REG_READY 0b00001111
+// CFG_REG_A
+#define COMP_TEMP_EN    (1 << 7)
+#define MD_CONTINUOUS   (0 << 0)
+#define ODR_100         ((1 << 3) | (1 << 2))
+// CFG_REG_B
+#define OFF_CANC        (1 << 1)
+// CFG_REG_C
+#define BDU             (1 << 4)
+
+static bool lis2mdlInit(magDev_t *mag)
+{
+    bool ack = true;
+    extDevice_t *dev = &mag->dev;
+
+    busDeviceRegister(dev);
+
+    ack = ack && busWriteRegister(dev, LIS2MDL_ADDR_CFG_REG_A, MD_CONTINUOUS | ODR_100 | COMP_TEMP_EN);
+    ack = ack && busWriteRegister(dev, LIS2MDL_ADDR_CFG_REG_B, OFF_CANC);
+    ack = ack && busWriteRegister(dev, LIS2MDL_ADDR_CFG_REG_C, BDU);
+
+    if (!ack) {
+        return false;
+    }
+
+    mag->magOdrHz = 100;
+    return true;
+}
+
+static bool lis2mdlRead(magDev_t *mag, int16_t *magData)
 {
-    static uint8_t buf[6];
-    static bool pendingRead = true;
+    uint8_t status = 0;
+    uint8_t buf[6] = {};
```

Comment:
You don't need to zero-init it
```suggestion
    uint8_t buf[6];
```